scripts/generate-changelog.ts

```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright 2017 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Generates entries for CHANGELOG.md from pull requests
+ *
+ * Reads changelog entries from pull requests merged since the last release tag.
+ * Changelog entries are lines within the first PR comment that matches /^\[[a-z\-]+\]/ like `[new-rule]` and `[bugfix]`
+ */
+
+// tslint:disable:no-console
+
+import GitHubApi = require("github");
+import * as fs from "fs";
+
+const github = new GitHubApi({
+    host: "api.github.com",
+    protocol: "https",
+    timeout: 5000,
+});
+
+const repoInfo = {
+    owner: "palantir",
+    repo: "tslint",
+};
+
+// authenticate
+const auth: GitHubApi.Auth = {
+    token: fs.readFileSync("../../github_token.txt", "utf8").toString().trim(),
+    type: "oauth",
+};
+console.log("Using OAuth token " + auth.token + "\n");
+github.authenticate(auth);
+
+const commits: ICommit[] = [];
+github.repos.getLatestRelease(repoInfo).then((value) => {
+    console.log("Getting commits " + value.tag_name + "..master");
+    // get the commits between the most recent release and the head of master
+    return github.repos.compareCommits({
+        base: value.tag_name,
+        head: "master",
+        ...repoInfo,
+    });
+}).then((value) => {
+    // for each commit, get the PR, and extract changelog entries
+    const promises: Array<Promise<any>> = [];
+    for (const commitInfo of value.commits) {
+        const commit: ICommit = {
+            fields: [],
+            sha: commitInfo.sha,
+            submitter: commitInfo.commit.author.name != null ? commitInfo.commit.author.name : commitInfo.author.login,
+            title: commitInfo.commit.message,
+        };
+        commits.push(commit);
+
+        // check for a pull request number in the commit title
+        const match = (commitInfo.commit.message as string).match(/\(#(\d+)\)/);
+        if (match && match.length > 1) {
+            commit.pushRequestNum = Number.parseInt(match[1], 10);
+
+            // get the PR text
+            promises.push(github.issues.get({
+                number: commit.pushRequestNum,
+                ...repoInfo,
+            }).then((comment) => {
+                // extract the changelog entries
+                const lines = (comment.body as string).split("\r\n");
+                for (const line of lines) {
+                    const fieldMatch = line.match(/^(\[[a-z\-]+\])/);
```

Comment:
we maybe should have a pre-defined list of tags, and users will have to pick from one of them